### PR TITLE
Operation with native integer constant that might overflow in unchecked context should be treated as not constant

### PR DESCRIPTION
**Pull request by cston**
For conversions that depend on the platform, the conversion should be treated as not constant and computed at runtime rather than compile-time.

Fixes https://github.com/dotnet/roslyn/issues/42955
Fixes https://github.com/dotnet/roslyn/issues/45525
Fixes https://github.com/dotnet/roslyn/issues/45531

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -6283,4 +6283,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_SealedGetHashCodeInRecord" xml:space="preserve">
     <value>'{0}' cannot be sealed because containing 'record' is not sealed.</value>
   </data>
+  <data name="WRN_ConstOutOfRangeChecked" xml:space="preserve">
+    <value>Constant value '{0}' may overflow '{1}' (use 'unchecked' syntax to override)</value>
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6287]: >may overflow [](start = 32, length = 12)

Do we want to add "at run-time"?

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 2]
```diff
@@ -1607,14 +1624,22 @@ private static bool CheckConstantBounds(SpecialType destinationType, double valu
...
             return true;
         }
 
-        private static bool CheckConstantBounds(SpecialType destinationType, decimal value)
+        private static bool CheckConstantBounds(SpecialType destinationType, decimal value, out bool maySucceedAtRuntime)
         {
-            // Dev10 checks (minValue - 1) < value < (MaxValue + 1) + 1).
+            maySucceedAtRuntime = false;
+
+            // Dev10 checks (minValue - 1) < value < (maxValue + 1).
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1642]: >maxValue [](start = 54, length = 8)

Does this change accurately reflect what Dev10 does?

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 3]
```diff
@@ -1211,7 +1211,7 @@ static ErrorCode getRefMismatchErrorCode(TypeKind type)
             return null;
         }
 
-        private ConstantValue FoldConstantNumericConversion(
+        private ConstantValue? FoldConstantNumericConversion(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1214]: >FoldConstantNumericConversion [](start = 31, length = 29)

Are all existing consumers prepared able to deal with null as the result?

Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:1214]: Yes, although I've added a few more annotations and handled `null` in one case that is currently obscured by [https://github.com/dotnet/roslyn/issues/45625](https://github.com/dotnet/roslyn/issues/45625).

---
In reply to: [449108702](https://github.com/dotnet/roslyn/pull/45581#discussion_r449108702) [](ancestors = 449108702)

**Comment by chsienki** [comment]
Latest changes LGTM

**Pull request merged**